stan/math/prim/mat/prob/ordered_logistic_lpmf.hpp

```diff
@@ -46,55 +51,196 @@ namespace stan {
+                          const Eigen::Matrix<T_cut, Eigen::Dynamic, 1>& c) {
+      typename return_type<T_loc>::type logp_n(0.0);
+
+      using std::exp;
+      using std::log;
 
+      static const std::string function = "ordered_logistic";
+
+      int K = c.size() + 1;
+
+      check_bounded(function, "Random variable", y, 1, K);
+      check_finite(function, "Location parameter", lambda);
+      check_greater(function, "Size of cut points parameter", c.size(), 0);
+      check_ordered(function, "Cut-points", c);
+      check_finite(function, "Final cut-point", c(c.size()-1));
+      check_finite(function, "First cut-point", c(0));
+
+      if (y == 1) {
+        logp_n -= log1p_exp(lambda - c[0]);
+      } else if (y == K) {
+        logp_n -= log1p_exp(c[K-2] - lambda);
+      } else {
+        logp_n += log_inv_logit_diff(c[y-2] - lambda, c[y-1] - lambda);
+      }
+      return logp_n;
+    }
+
+    template <typename T_loc, typename T_cut>
+    typename return_type<T_loc>::type
+    ordered_logistic_lpmf(int y, const T_loc& lambda,
+                          const Eigen::Matrix<T_cut, Eigen::Dynamic, 1>& c)  {
+      return ordered_logistic_lpmf<false>(y, lambda, c);
+    }
+
+    /**
+     * Returns the (natural) log probability of the specified array 
+     * of integers given the vector of continuous locations and 
+     * specified cutpoints in an ordered logistic model.
+     *
+     * <p>Typically the continous location
+     * will be the dot product of a vector of regression coefficients
+     * and a vector of predictors for the outcome.
+     *
+     * @tparam propto True if calculating up to a proportion.
+     * @tparam T_loc Location type.
+     * @tparam T_cut Cut-point type.
+     * @param y Array of integers
+     * @param lambda Vector of continuous location variables.
+     * @param c Positive increasing vector of cutpoints.
+     * @return Log probability of outcome given location and
+     * cutpoints.
      * @throw std::domain_error If the outcome is not between 1 and
      * the number of cutpoints plus 2; if the cutpoint vector is
      * empty; if the cutpoint vector contains a non-positive,
      * non-finite value; or if the cutpoint vector is not sorted in
      * ascending order.
+     * @throw std::invalid_argument If y and lambda are different
+     * lengths.
      */
-    template <bool propto, typename T_lambda, typename T_cut>
-    typename boost::math::tools::promote_args<T_lambda, T_cut>::type
-    ordered_logistic_lpmf(int y, const T_lambda& lambda,
-                         const Eigen::Matrix<T_cut, Eigen::Dynamic, 1>& c) {
+    template <bool propto, typename T_loc, typename T_cut>
+    typename return_type<T_loc>::type
+    ordered_logistic_lpmf(const std::vector<int>& y,
+                          const Eigen::Matrix<T_loc, Eigen::Dynamic, 1>& lambda,
+                          const Eigen::Matrix<T_cut, Eigen::Dynamic, 1>& c) {
+      typename return_type<T_loc>::type logp_n(0.0);
```

Comment:
As a matter of style, you should declare variables as far down as you can so that their first use is as close to the declaration as possible.  Here that'd mean right before the for loop.  This also saves doing an allocation in case you throw, but that's not an issue here since throwing will clean everything anyway.

Again, no need to change here, just a heads up for style in the future.